File: mobile/lib/features/workout/domain/workout_session.dart
// =====================================================================
// MiCoach — Sesión de entrenamiento (historial), mirror de WorkoutDtos.
// =====================================================================
class SessionExercise {
  final int id;
  final int? workoutExerciseId;
  final int exerciseId;
  final int? setsDone;
  final double? weightKg;
  final int? reps;
  final int? rpe;
  final int? durationSeconds;
  final int? distanceMeters;
  final String? notes;

  const SessionExercise({
    required this.id,
    this.workoutExerciseId,
    required this.exerciseId,
    this.setsDone,
    this.weightKg,
    this.reps,
    this.rpe,
    this.durationSeconds,
    this.distanceMeters,
    this.notes,
  });

  factory SessionExercise.fromJson(Map<String, dynamic> json) => SessionExercise(
        id: json['id'] as int,
        workoutExerciseId: json['workoutExerciseId'] as int?,
        exerciseId: json['exerciseId'] as int,
        setsDone: json['setsDone'] as int?,
        weightKg: (json['weightKg'] as num?)?.toDouble(),
        reps: json['reps'] as int?,
        rpe: json['rpe'] as int?,
        durationSeconds: json['durationSeconds'] as int?,
        distanceMeters: json['distanceMeters'] as int?,
        notes: json['notes'] as String?,
      );
}

class WorkoutSession {
  final int id;
  final int? workoutId;
  final int? workoutDayId;
  final String status;
  final DateTime? startedAt;
  final DateTime? completedAt;
  final int? durationSeconds;
  final String? notes;
  final List<SessionExercise> exercises;

  const WorkoutSession({
    required this.id,
    this.workoutId,
    this.workoutDayId,
    required this.status,
    this.startedAt,
    this.completedAt,
    this.durationSeconds,
    this.notes,
    this.exercises = const [],
  });

  factory WorkoutSession.fromJson(Map<String, dynamic> json) => WorkoutSession(
        id: json['id'] as int,
        workoutId: json['workoutId'] as int?,
        workoutDayId: json['workoutDayId'] as int?,
        status: json['status'] as String,
        startedAt: json['startedAt'] != null ? DateTime.parse(json['startedAt'] as String) : null,
        completedAt: json['completedAt'] != null ? DateTime.parse(json['completedAt'] as String) : null,
        durationSeconds: json['durationSeconds'] as int?,
        notes: json['notes'] as String?,
        exercises: (json['exercises'] as List<dynamic>? ?? const [])
            .map((e) => SessionExercise.fromJson(e as Map<String, dynamic>))
            .toList(),
      );
}

File: backend/modules/workout/src/main/java/com/micoach/workout/domain/WorkoutSession.java
package com.micoach.workout.domain;

import lombok.Getter;

import java.time.Instant;
import java.util.List;

/**
 * Sesión de entrenamiento realizada, historial (tabla workout_sessions).
 */
@Getter
public class WorkoutSession {

    private final Long id;
    private final Long userId;
    private final Long workoutId;
    private final Long workoutDayId;
    private String status;
    private final Instant startedAt;
    private Instant completedAt;
    private Integer durationSeconds;
    private String notes;
    private List<SessionExercise> exercises;
    private final Instant createdAt;
    private Instant updatedAt;

    private WorkoutSession(Long id, Long userId, Long workoutId, Long workoutDayId, String status,
                           Instant startedAt, Instant completedAt, Integer durationSeconds,
                           String notes, List<SessionExercise> exercises, Instant createdAt,
                           Instant updatedAt) {
        this.id = id;
        this.userId = userId;
        this.workoutId = workoutId;
        this.workoutDayId = workoutDayId;
        this.status = status;
        this.startedAt = startedAt;
        this.completedAt = completedAt;
        this.durationSeconds = durationSeconds;
        this.notes = notes;
        this.exercises = exercises == null ? List.of() : exercises;
        this.createdAt = createdAt;
        this.updatedAt = updatedAt;
    }

    public static WorkoutSession start(Long userId, Long workoutId, Long workoutDayId) {
        Instant now = Instant.now();
        return new WorkoutSession(null, userId, workoutId, workoutDayId, "in_progress", now,
                null, null, null, List.of(), now, now);
    }

    public static WorkoutSession restore(Long id, Long userId, Long workoutId, Long workoutDayId,
                                         String status, Instant startedAt, Instant completedAt,
                                         Integer durationSeconds, String notes,
                                         List<SessionExercise> exercises, Instant createdAt,
                                         Instant updatedAt) {
        return new WorkoutSession(id, userId, workoutId, workoutDayId, status, startedAt,
                completedAt, durationSeconds, notes, exercises, createdAt, updatedAt);
    }

    public void complete(Integer durationSeconds, String notes) {
        this.status = "completed";
        this.completedAt = Instant.now();
        this.durationSeconds = durationSeconds;
        this.notes = notes;
        this.updatedAt = Instant.now();
    }

    public void abort(String notes) {
        this.status = "aborted";
        this.completedAt = Instant.now();
        this.notes = notes;
        this.updatedAt = Instant.now();
    }

    public boolean belongsTo(Long userId) {
        return this.userId.equals(userId);
    }
}
